### PR TITLE
CHET-169: Add state machine hooks into database migration service

### PR DESCRIPTION
**Pull request by theghostwhoforks**
wanted to get @tarka's initial thoughts on database migration status and the need, if any, to co-exist with the state machine

### CODE REVIEW

**Comment by theghostwhoforks** [comment]
I can't test if the beans are being injected without the java osgi refactor branch being merged in.

Once that is updated, I'll test and post the results here

File: src/test/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseArchivalServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        service = new DatabaseArchivalService(migrationService, databaseExtractor);
+    }
+
+    @Test
+    void shouldArchiveDatabaseSuccessfully() throws Exception {
+        when(this.databaseExtractor.startDatabaseDump(tempDir.resolve("db.dump"))).thenReturn(process);
+        when(process.waitFor()).thenReturn(0);
+        Path target = service.archiveDatabase(tempDir);
+        assertTrue(target.endsWith("db.dump"));
+        //TODO: Should this stage to transition to be called pending export?
```
Comment (bpar476) [inline src/test/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseArchivalServiceTest.java]: I think the current nomenclature is fine.

File: src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java [hunk 2]
```diff
@@ -1,89 +1,70 @@
...
-            throw new DatabaseMigrationFailure(msg, e);
-        }
-        setStatus(MigrationStatus.DUMP_COMPLETE);
-
+    public FileSystemMigrationErrorReport performMigration() throws DatabaseMigrationFailure, InvalidMigrationStageError {
+        Path pathToDatabaseFile = databaseArchivalService.archiveDatabase(tempDirectory);
+        return uploadDatabaseArtifactToS3(pathToDatabaseFile);
+    }
 
+    //TODO: This should be decomposed as a standalone service for easier testing
```
Comment (bpar476) [inline src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java]: I think I agree with this. If the database dumping is being separated from the service then the storage probably should as well.

File: src/main/java/com/atlassian/migration/datacenter/spi/MigrationStage.java [hunk 3]
```diff
@@ -8,13 +8,24 @@
...
+
+    WAIT_PROVISION_MIGRATION_STACK("wait_provision_migration"),
+    PROVISION_MIGRATION_STACK("provision_migration"),
+    WAIT_PROVISION_APPLICATION("wait_provision_app"),
+    PROVISION_APPLICATION("provision_app"),
     AUTHENTICATION("authentication"),
-    NOT_STARTED("");
+    NOT_STARTED("not_started");
 
     private String key;
```
Comment (tarka) [inline src/main/java/com/atlassian/migration/datacenter/spi/MigrationStage.java]: What's the purpose of this?

File: src/main/java/com/atlassian/migration/datacenter/spi/MigrationStage.java [hunk 4]
```diff
@@ -8,13 +8,24 @@
     FINISHED("finished"),
     CUTOVER("cutover"),
     VALIDATE("validate"),
-    DB_MIGRATION_IMPORT("db_migration_down"), DB_MIGRATION_EXPORT("db_migration_up"),
+
+    DB_MIGRATION_IMPORT("db_migration_down"),
```
Comment (tarka) [inline src/main/java/com/atlassian/migration/datacenter/spi/MigrationStage.java]: I'm replace the DB migration state enum with this one. However I if we have an error we should include more information; I think a wrapper class that can encapsulate this information should work.

File: src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java (hunk 2 above)
Comment (theghostwhoforks) [reply src/main/java/com/atlassian/migration/datacenter/core/aws/db/DatabaseMigrationService.java]: Cool. Will wrap this in a small service

File: src/main/java/com/atlassian/migration/datacenter/spi/MigrationStage.java (hunk 3 above)
Comment (theghostwhoforks) [reply src/main/java/com/atlassian/migration/datacenter/spi/MigrationStage.java]: This is here as part of the `DevelopEndpoint`. This was added to aid in deserialization of a `MigrationStage` object

cc: @nanux

File: src/main/java/com/atlassian/migration/datacenter/spi/MigrationStage.java (hunk 4 above)
Comment (theghostwhoforks) [reply src/main/java/com/atlassian/migration/datacenter/spi/MigrationStage.java]: I've updated all the enums to represent the state transitions.

/nit. I'd be a tad happier if we can make the currentStage return a callback for `next()` transition that accepts a migration service 

Something like this -
`getCurrentStage().next().apply(migrationService)`
can be accomplished easily.

I'll leave this for another refactoring.

**Comment by theghostwhoforks** [comment]
Note to self:
This is pending -
1. Fix Develop endpoint 
2. Extract a utility function to a common builder

Once done, and tested, this PR will be made ready to merge

File: src/main/java/com/atlassian/migration/datacenter/core/aws/infrastructure/QuickstartDeploymentService.java [hunk 5]
```diff
@@ -60,7 +60,7 @@ public QuickstartDeploymentService(ActiveObjects ao, CfnApi cfnApi, MigrationSer
     @Override
     public void deployApplication(String deploymentId, Map<String, String> params) throws InvalidMigrationStageError {
         logger.info("received request to deploy application");
-        migrationService.transition(MigrationStage.PROVISION_APPLICATION, MigrationStage.WAIT_PROVISION_APPLICATION);
+        migrationService.transition(MigrationStage.PROVISION_APPLICATION_WAIT);
```
Comment (tarka) [inline src/main/java/com/atlassian/migration/datacenter/core/aws/infrastructure/QuickstartDeploymentService.java]: As we discussed in standup, we should probably pick one system. But we don't need to do it straight away if we want to see what approach works best.

**Comment by theghostwhoforks** [comment]
All merge conflicts have now been resolved. They were all logical conflicts, something that couldn't be automatically resolved. 

Once the tests pass, this shall be merged in ASAP

**Pull request merged**